adapters/valueimpression/valueimpression.go

```diff
@@ -0,0 +1,161 @@
+package valueimpression
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type Adapter struct {
+	endpoint string
+}
+
+func (a *Adapter) MakeRequests(request *openrtb.BidRequest, unused *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var errs []error
+	var adapterRequests []*adapters.RequestData
+
+	if err := preprocess(request); err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	adapterReq, err := a.makeRequest(request)
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	adapterRequests = append(adapterRequests, adapterReq)
+
+	return adapterRequests, errs
+}
+
+func (a *Adapter) makeRequest(request *openrtb.BidRequest) (*adapters.RequestData, error) {
+	var err error
+
+	jsonBody, err := json.Marshal(request)
+	if err != nil {
+		return nil, err
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    jsonBody,
+		Headers: headers,
+	}, nil
+}
+
+func preprocess(request *openrtb.BidRequest) error {
+	if len(request.Imp) == 0 {
+		return &errortypes.BadInput{
+			Message: "No Imps in Bid Request",
+		}
+	}
+	for i := 0; i < len(request.Imp); i++ {
+		var imp = &request.Imp[i]
+		var bidderExt adapters.ExtImpBidder
+
+		if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+			return &errortypes.BadInput{
+				Message: err.Error(),
+			}
+		}
+
+		if err := validateImp(imp); err != nil {
+			return err
+		}
+
+		var extImp openrtb_ext.ExtImpValueImpression
+		if err := json.Unmarshal(bidderExt.Bidder, &extImp); err != nil {
+			return &errortypes.BadInput{
+				Message: err.Error(),
+			}
+		}
+
+		imp.Ext = bidderExt.Bidder
+	}
+
+	return nil
+}
+
+func validateImp(imp *openrtb.Imp) error {
+	if imp.Banner == nil && imp.Video == nil {
+		return &errortypes.BadInput{
+			Message: "Only Banner and Video bid-types are supported at this time",
+		}
+	}
+	return nil
+}
+
+// MakeBids based on valueimpression server response
+func (a *Adapter) MakeBids(bidRequest *openrtb.BidRequest, unused *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if responseData.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
```

Comment:
It's also common to check for a `StatusBadRequest` return type and map it to a `errortypes.BadInput`.